File: frontend/src/services/assessmentService.ts
import { apiClient, ENDPOINTS } from '../api';
import type {
    Assessment,
    AssessmentAttempt,
    AssessmentAttemptRequest,
    AssessmentRequest,
    QuestionRequest
} from '../types';

export const assessmentService = {
    getByCourse: async (courseId: string): Promise<Assessment[]> => {
        return await apiClient.get<Assessment[]>(
            ENDPOINTS.ASSESSMENTS.BY_COURSE(courseId)
        );
    },

    getById: async (id: string): Promise<Assessment> => {
        return await apiClient.get<Assessment>(
            ENDPOINTS.ASSESSMENTS.BY_ID(id)
        );
    },

    create: async (data: AssessmentRequest): Promise<Assessment> => {
        return await apiClient.post<Assessment>(
            ENDPOINTS.ASSESSMENTS.BASE,
            data
        );
    },

    addQuestion: async (data: QuestionRequest): Promise<Assessment> => {
        return await apiClient.post<Assessment>(
            ENDPOINTS.ASSESSMENTS.QUESTIONS,
            data
        );
    },

    deleteQuestion: async (questionId: string): Promise<void> => {
        await apiClient.delete<void>(
            ENDPOINTS.ASSESSMENTS.DELETE_QUESTION(questionId)
        );
    },

    startAttempt: async (assessmentId: string): Promise<AssessmentAttempt> => {
        return await apiClient.post<AssessmentAttempt>(
            ENDPOINTS.ASSESSMENTS.START(assessmentId)
        );
    },

    submitAttempt: async (data: AssessmentAttemptRequest): Promise<AssessmentAttempt> => {
        return await apiClient.post<AssessmentAttempt>(
            ENDPOINTS.ASSESSMENTS.SUBMIT,
            data
        );
    },

    getMyAttempts: async (): Promise<AssessmentAttempt[]> => {
        return await apiClient.get<AssessmentAttempt[]>(
            ENDPOINTS.ASSESSMENTS.MY_ATTEMPTS
        );
    },

    getAttempt: async (attemptId: string): Promise<AssessmentAttempt> => {
        return await apiClient.get<AssessmentAttempt>(
            ENDPOINTS.ASSESSMENTS.ATTEMPT_BY_ID(attemptId)
        );
    },

    hasCompletedPretest: async (courseId: string): Promise<boolean> => {
        return await apiClient.get<boolean>(
            ENDPOINTS.ASSESSMENTS.CHECK_PRETEST(courseId)
        );
    },

    hasCompletedPosttest: async (courseId: string): Promise<boolean> => {
        return await apiClient.get<boolean>(
            ENDPOINTS.ASSESSMENTS.CHECK_POSTTEST(courseId)
        );
    },
};

File: backend/src/main/java/hr/algebra/adaptive/learning/backend/service/AssessmentService.java
package hr.algebra.adaptive.learning.backend.service;

import hr.algebra.adaptive.learning.backend.dto.assessment.*;

import java.util.List;
import java.util.UUID;

public interface AssessmentService {

    AssessmentResponse create(AssessmentRequest request, UUID createdById);

    AssessmentResponse getById(UUID id);

    AssessmentResponse getByIdWithQuestions(UUID id);

    List<AssessmentResponse> getByCourse(UUID courseId);

    AssessmentResponse addQuestion(QuestionRequest request);

    void deleteQuestion(UUID questionId);

    AssessmentAttemptResponse startAttempt(UUID assessmentId, UUID studentId);

    AssessmentAttemptResponse submitAttempt(AssessmentAttemptRequest request, UUID studentId);

    AssessmentAttemptResponse getAttempt(UUID attemptId);

    List<AssessmentAttemptResponse> getStudentAttempts(UUID studentId);

    boolean hasCompletedPretest(UUID studentId, UUID courseId);

    boolean hasCompletedPosttest(UUID studentId, UUID courseId);
}
